File: r/decision_tree.R
# ==============================================
# 🌳 Decision Tree Classification (R)
# ==============================================

# 1. Load Dataset
dataset <- read.csv('../data/social_network_ads.csv')

# 2. Split into Training and Test Sets
library(caTools)
set.seed(123)
split           <- sample.split(dataset$Purchased, SplitRatio = 0.8)
train_set       <- subset(dataset, split == TRUE)
test_set        <- subset(dataset, split == FALSE)

# 3. Retain Unscaled Copies for Tree Visualization
train_unscaled  <- train_set
test_unscaled   <- test_set

# 4. Feature Scaling (for decision boundary plots)
train_set[, 1:2] <- scale(train_set[, 1:2])
test_set[,  1:2] <- scale(test_set[,  1:2])

# 5. Train Decision Tree Models
library(rpart)

# 5a. Unscaled model (for plotting tree structure)
tree_unscaled <- rpart(
  Purchased ~ .,
  data = train_unscaled,
  
)

# 5b. Scaled model (for decision region)
tree_scaled   <- rpart(
  Purchased ~ .,
  data = train_set,
  
)

# 6. Predict on Test Set (scaled)
y_pred <- as.numeric(predict(tree_scaled, newdata = test_set) > 0.5)

# 7. Confusion Matrix
cm <- table(
  Actual    = test_set$Purchased,
  Predicted = y_pred
)
print(cm)

# 8. Plot the Tree Structure (unscaled)
plot(tree_unscaled, main = "Decision Tree Structure (Unscaled Data)")
text(tree_unscaled, use.n = TRUE, all = TRUE, cex = 0.8)

# 9. Decision Boundary Plot Function
plot_decision_boundary <- function(data, title) {
  X1 <- seq(min(data$Age) - 1, max(data$Age) + 1, by = 0.01)
  X2 <- seq(min(data$EstimatedSalary) - 1, max(data$EstimatedSalary) + 1, by = 0.01)
  grid <- expand.grid(Age = X1, EstimatedSalary = X2)
  
  # Predict and threshold
  preds <- as.numeric(predict(tree_scaled, newdata = grid) > 0.5)
  
  # Base plot
  plot(
    data$Age, data$EstimatedSalary,
    type = 'n', main = title,
    xlab = 'Age (scaled)', ylab = 'Estimated Salary (scaled)',
    xlim = range(X1), ylim = range(X2)
  )
  contour(X1, X2, matrix(preds, length(X1), length(X2)), add = TRUE)
  
  # Plot decision regions
  points(
    grid, pch = '.', 
    col = ifelse(preds == 1, 'springgreen3', 'tomato'),
    cex = 0.5
  )
  # Plot true points
  points(
    data$Age, data$EstimatedSalary,
    pch = 21, bg = ifelse(data$Purchased == 1, 'green4', 'red3')
  )
}

# 10. Visualize Decision Regions
plot_decision_boundary(train_set, "Decision Tree (Training Set)")
plot_decision_boundary(test_set,  "Decision Tree (Test Set)")

File: python/decision_tree.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 26 20:34:22 2025
Author: Chukwuka Chijioke Jerry

Decision Tree Classification on Social Network Ads
"""

# ===========================
# 📦 Import Libraries
# ===========================
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, accuracy_score
from matplotlib.colors import ListedColormap
import seaborn as sns

# ===========================
# 📂 Load Dataset
# ===========================
dataset = pd.read_csv('../data/social_network_ads.csv')
X = dataset.iloc[:, :-1].values  # Age, EstimatedSalary
y = dataset.iloc[:, -1].values   # Purchased

# ===========================
# 🔀 Train/Test Split
# ===========================
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# ===========================
# ⚖️ Feature Scaling (for visualization)
# ===========================
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test  = scaler.transform(X_test)

# ===========================
# 🌳 Train Decision Tree Model
# ===========================
classifier = DecisionTreeClassifier(random_state=0)
classifier.fit(X_train, y_train)

# ===========================
# 🔍 Predict & Evaluate
# ===========================
y_pred = classifier.predict(X_test)

# Confusion Matrix & Accuracy
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)

print("Confusion Matrix:")
print(cm)
print(f"Accuracy: {accuracy:.2f}")

sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False,
            xticklabels=['Not Purchased','Purchased'],
            yticklabels=['Not Purchased','Purchased'])
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.title('Confusion Matrix')
plt.tight_layout()
plt.show()

# ===========================
# 📊 Decision Boundary Plotter
# ===========================
def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.arange(X_set[:,0].min() - 1, X_set[:,0].max() + 1, 0.01),
        np.arange(X_set[:,1].min() - 1, X_set[:,1].max() + 1, 0.01)
    )
    Z = classifier.predict(np.c_[X1.ravel(), X2.ravel()]).reshape(X1.shape)

    plt.contourf(
        X1, X2, Z,
        alpha=0.3,
        cmap=ListedColormap(('red','green'))
    )
    plt.xlim(X1.min(), X1.max())
    plt.ylim(X2.min(), X2.max())

    for idx, label in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == label, 0],
            X_set[y_set == label, 1],
            c=('red','green')[idx],
            label=f'Class {label}'
        )
    plt.title(title)
    plt.xlabel('Age (standardized)')
    plt.ylabel('Estimated Salary (standardized)')
    plt.legend()
     
    plt.tight_layout()
    plt.show()

# ===========================
# 🖼️ Visualize Training & Test Sets
# ===========================
plot_decision_boundary(X_train, y_train, 'Decision Tree (Training Set)')
plot_decision_boundary(X_test,  y_test,  'Decision Tree (Test Set)')
